### PR TITLE
Enhance "Custom dictionary style" sample to use DictionarySymbolStyle created from a web style

### PR DESCRIPTION
**Pull request by sclaridge**
### Description

This PR adds an enhancement to the "Custom dictionary style" sample to toggle between a `DictionarySymbolStyle` created from a web style and a local stylex file. The dictionary style created from a web style displays the health grade of the restaurants, so that the symbols are different when the user toggles between the two styles.

The README has been updated to reflect the new workflow and web style data.

### CODE REVIEW

File: symbology/custom-dictionary-style/README.md
Comment (TADraeseke) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
Use the radio buttons to toggle between the dictionary symbols from the web style and style file. Pan and zoom around the map to see the symbology from the chosen dictionary symbol style. The web style and style file are slightly different to each other to give a visual indication of the switch between the two.
```

Comment (TADraeseke) [reply symbology/custom-dictionary-style/README.md]: Bit of unsolicited review there :D. Not sure if the sentence I've add reads well, but just something to clue the user in to that expected difference when using one style vs the other

File: symbology/custom-dictionary-style/README.md
Comment (TADraeseke) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
2. Based on the style selected:
  -  If the web style toggle has been selected, create a new `DictionarySymbolStyle` from the portal item using `new DictionarySymbolStyle(portalItem)`
  -  If the file style toggle has been selected, create a new dictionary symbol style using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`
```

Just stepping out of the numbered ordering here since they're "ifs"

Comment (JonLavi) [reply symbology/custom-dictionary-style/README.md]: Agree with this. I also think we could perhaps incoroprate point 1 (creating the PortalItem`) into the respective 'if' here, as the portal item is only necessary for that particular action, not for the whole sample.

File: symbology/custom-dictionary-style/README.md
Comment (JonLavi) [inline symbology/custom-dictionary-style/README.md]: Maybe we can break this down into two paragraphs, one about the data, one about the renderer?

```suggestion
The data used in this sample is from a feature layer showing a subset of [restaurants in Redlands, CA](https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/arcgis/rest/services/Redlands_Restaurants/FeatureServer) hosted as a feature service with attributes for rating, style, health score, and open hours.

The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylex file](https://arcgisruntime.maps.arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgisruntime.maps.arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (JonLavi) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: I'm a bit rusty on my Java these days, but I think `Map` is an interface, which would be somewhat ambiguous to put an object of the type `HashMap` in it. Perhaps a better way to write this would be 

`HashMap<String, String> fieldMap = new HashMap<>();`

The method call on line 139 (`new DictionaryRenderer(dictionarySymbolStyle, fieldMap, fieldMap);`) which takes this variable in has a signature that takes a `Map` object, but since `HashMap` is a type of `Map`, it should be all good.

I'd happily stand corrected though!

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (JonLavi) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: I'm sure you've thought of this one before: I originally saw there might be a bit of scope for refactoring these three lines here, by adding them to the end of the statement `toggleGroup.selectedToggleProperty().addListener(changed -> {`
It would mean that the statements `if (toggleGroup.getSelectedToggle() == webStyleButton/fileStyleButton` would only be responsible for populating the `dictionaryRenderer`, and after that, the renderer/map/progress indicator get dealt with either way.

However, the whoe async process under `if (toggleGroup.getSelectedToggle() == webStyleButton) {` kind of complicates the whole thing! 
Can't think of a different pattern just now, and I don't think it's worth trying to come up with something less readable than what we've got at the moment.

```suggestion
      toggleGroup.selectedToggleProperty().addListener(changed -> {
        // clear the existing dictionary symbol style and feature layer on the map
        map.getOperationalLayers().clear();
        dictionarySymbolStyle = null;
        dictionaryRenderer = null;

        // turn on the progress indicator when the symbols are loading
        progressIndicator.setVisible(true);

		// create the renderer based on which source is selected for the symbol style
        if (toggleGroup.getSelectedToggle() == webStyleButton) {
          dictionarySymbolStyle = new DictionarySymbolStyle(portalItem);
          // load the symbol dictionary
          dictionarySymbolStyle.loadAsync();
          dictionarySymbolStyle.addDoneLoadingListener(() -> {
            if (dictionarySymbolStyle.getLoadStatus() == LoadStatus.LOADED) {
              // create a dictionary renderer with the dictionary symbol style
              dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle, fieldMap, fieldMap);
            } else {
              new Alert(Alert.AlertType.ERROR, "Dictionary symbol style failed to load!").show();
            }
          });
        }
        if (toggleGroup.getSelectedToggle() == fileStyleButton){
          // create a symbol dictionary from the local symbol style file
          dictionarySymbolStyle = DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath());
          // create a dictionary renderer with the dictionary symbol style
          dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle);
        }
        
        // set the renderer to the feature layer
        restaurantLayer.setRenderer(dictionaryRenderer);
        // add the layer to the map
        map.getOperationalLayers().add(restaurantLayer);
        // hide the progress indicator when the symbols have been applied to the features
        progressIndicator.setVisible(false);
      });
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (JonLavi) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: Perhaps we can add a line here to select one of the things by default? I found it a little odd that the sample loads up but none is selected, having to select one manually.

File: symbology/custom-dictionary-style/src/main/resources/custom_dictionary_style/style.css
Comment (JonLavi) [inline symbology/custom-dictionary-style/src/main/resources/custom_dictionary_style/style.css]: Missing a newline here 😁

**Review by JonLavi — APPROVED** [review]
Nice work @sclaridge !
This sample was quite a difficult one to make accessible back in it's first iteration, and this is quite a nice addition to it, keeping it clean and simple.
Just seconding @TADraeseke 's comments on the Readme. Other than that, a couple of observations regarding the code, but nothing needing changing. Going to approve so you can move on with a second review 👍

**Comment by sclaridge** [comment]
@Rachael-E I've edited the README based on Jon and Trevor's suggestions.   
  
The style file symbols are now displayed when the samples starts up using `oldToggle` and `newToggle` values from the `toggleGroup.selectedToggleProperty().addListener()`.

File: symbology/custom-dictionary-style/README.md
Comment (JonLavi) [inline symbology/custom-dictionary-style/README.md]: Looks like you can delete point 3 here, and update the numbers further down afterwards 😄

File: symbology/custom-dictionary-style/README.md
Comment (Rachael-E) [inline symbology/custom-dictionary-style/README.md]: we can simplify this step a bit: keeping the focus on the dictionary symbol functionality without explicitly spelling out the loading API pattern. 
```suggestion
3. Load the dictionary symbol style. Once it has loaded, create a new `DictionaryRenderer`, providing the dictionary symbol style.
```

File: symbology/custom-dictionary-style/README.md
Comment (Rachael-E) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
-  If the file style toggle has been selected, create a new `DictionarySymbolStyle` using `DictionarySymbolStyle.createFromFile(stylxFile.getAbsolutePath())`
```

File: symbology/custom-dictionary-style/README.md
Comment (Rachael-E) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
6. Add the feature layer to the map's operational layers using `getOperationalLayers().add(featureLayer)`.
```

File: symbology/custom-dictionary-style/README.md
Comment (Rachael-E) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
5. Apply the dictionary renderer to a feature layer using `featureLayer.setRenderer(dictionaryRenderer)`.
```

File: symbology/custom-dictionary-style/README.md
Comment (Rachael-E) [inline symbology/custom-dictionary-style/README.md]: we should have mention here of the other constructor too: the one with the field overrides (it's used for the web style option: any reason why and we can explain that to the user here?

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: let's split this up over two lines for readability
```suggestion
      map.setInitialViewpoint(
      new Viewpoint(new Point(-1.304630524635E7, 4036698.1412000023, map.getSpatialReference()), 5000));
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]:
```suggestion
      // enable UI interactions once the map has loaded
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: it looks like we don't need to set these to null since whenever the toggle is switched, it creates a new DictionarySymbolStyle/DictionaryRenderer object.
```suggestion
 
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: we should expand this comment to explain this constructor. e.g. why are we passing in the field map twice, what does it do in the sample? why is it important to the web style but not the stylx file?

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: let's avoid the use of "tells", it's a bit too informal for doc language: how about the following

```suggestion
              // set the dictionary renderer as the feature layer's renderer to apply symbols to features
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: same as above, let's avoid "tells"

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]:
```suggestion
      // add the control panel, map view, and progress indicator to the stack pane
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: this doesn't have to be a member variable: let's declare it locally since each time a new DictionaryRenderer object is being created anyway.

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: local instantiation here is fine: and we can make use of `var` type inference too since it's a lengthy class declaration but the variable naming is clear what the type is.
```suggestion
              var dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle, fieldMap, fieldMap);
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]:
```suggestion
          var dictionaryRenderer = new DictionaryRenderer(dictionarySymbolStyle);
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: I'm keen to get us thinking about using type inference more often to improve readability. this is a good example: we can use var and also split this over two lines and it becomes very readable:

```suggestion
      var serviceFeatureTable = 
      new ServiceFeatureTable("https://services2.arcgis.com/ZQgQTuoyBrtmoGdP/ArcGIS/rest/services/Redlands_Restaurants/FeatureServer/0");
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (Rachael-E) [inline symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: we wouldn't use `var` here since we've named this variable `restaurantLayer`: just highlighting examples and working out best practices here!

**Review by Rachael-E — CHANGES_REQUESTED** [review]
thanks @sclaridge ! a few things to consider here and we'll have a chat about a refactor I have in mind that will tidy up some of the repeated code here when you have a moment.

File: symbology/custom-dictionary-style/README.md
Comment (sclaridge) [reply symbology/custom-dictionary-style/README.md]: The field overrides are used so that the symbols created from the web style are different from the symbols automatically rendered from the style file. 
```suggestion
4. To contrast the symbols rendered from the two `DictionarySymbolStyle`s, create a new `DictionaryRenderer` providing the dictionary symbol style, and manually map the attributes from the feature layer to those expected by the `DictionarySymbolStyle`.
```

File: symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java
Comment (sclaridge) [reply symbology/custom-dictionary-style/src/main/java/com/esri/samples/custom_dictionary_style/CustomDictionaryStyleSample.java]: Just wanted to confirm based on our refactoring chat, the `dictSymbStyleFromPortal` and `webStyleDictionaryRenderer` will be member variables to prevent them from being garbage collected.

**Comment by sclaridge** [comment]
@Rachael-E I've added the refactored code that we discussed from our chat. I've also used `var` for `dictRendFromFile`, let me know if that is used intuitively based on the newly refactored code.   
  
I have added comments and edited the README to explain why the `webStyleDictionaryRenderer` is created by manually mapping the attributes to the symbols in the dictionary symbol style created from the portal item.

**Review by Rachael-E — APPROVED** [review]
great job @sclaridge , this is looks great! I've added a small commit, it just updates the readme every so slightly: I liked your comment in the code that explains the input fields more so I think we're ok to leave that out the readme. I also reinstated the type for DictionaryRenderer since I think the variable name isn't too clear when it's shortened, so we can leave it in for readable. Great job!

File: symbology/custom-dictionary-style/README.md
Comment (yo1995) [inline symbology/custom-dictionary-style/README.md]:
```suggestion
The feature layer is symbolized using a dictionary renderer that displays a single symbol for all of these variables. The renderer uses symbols from a custom restaurant dictionary style created from a [stylx file](https://arcgis.com/home/item.html?id=751138a2e0844e06853522d54103222a) and a [web style](https://arcgis.com/home/item.html?id=adee951477014ec68d7cf0ea0579c800), available as an items from ArcGIS Online, to show unique symbols based on several feature attributes. The symbols it contains were created using ArcGIS Pro. The logic used to apply the symbols comes from an Arcade script embedded in the stylx file (which is a SQLite database), along with a JSON string that defines expected attribute names and configuration properties.
```

Comment (yo1995) [reply symbology/custom-dictionary-style/README.md]: Ha I thought I was on devtopia... Will make the same suggestion there! 😅

**Comment by sclaridge** [comment]
@Rachael-E I've applied some changes based on feedback from the common samples design PR. One last look over would be helpful before I merge, thanks!

**Pull request merged**
